app/backend/src/couchers/rate_limit.py

```diff
@@ -0,0 +1,214 @@
+    FRIEND_REQUEST_WARNING_LIMIT,
+    HOST_REQUEST_HARD_LIMIT,
+    HOST_REQUEST_WARNING_LIMIT,
+    RATE_LIMIT_INTERVAL,
+)
+from couchers.models import (
+    Conversation,
+    FriendRelationship,
+    GroupChat,
+    HostRequest,
+    RateLimitAction,
+    RateLimitViolation,
+    User,
+)
+from couchers.tasks import send_rate_limit_violation_report_email
+from couchers.utils import now
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+
+logger = logging.getLogger(__name__)
+
+
+def _get_user_host_requests_in_past_time_interval(session, user_id) -> list[dict]:
+    return (
+        session.execute(
+            select(
+                Conversation.created.label("created"),
+                HostRequest.host_user_id.label("host id"),
+                User.username.label("host username"),
+                User.city.label("host city"),
+            )
+            .join(Conversation, HostRequest.conversation_id == Conversation.id)
+            .join(User, HostRequest.host_user_id == User.id)
+            .where(HostRequest.surfer_user_id == user_id)
+            .where(Conversation.created >= now() - RATE_LIMIT_INTERVAL)
+        )
+        .mappings()
+        .all()
+    )
+
+
+def _get_user_friend_requests_in_past_time_interval(session, user_id) -> list[dict]:
+    return (
+        session.execute(
+            select(
+                FriendRelationship.time_sent,
+                User.id.label("to_user (ID)"),
+                User.username.label("to_user (username)"),
+                FriendRelationship.status,
+            )
+            .join(User, FriendRelationship.to_user_id == User.id)
+            .where(FriendRelationship.from_user_id == user_id)
+            .where(FriendRelationship.time_sent >= now() - RATE_LIMIT_INTERVAL)
+        )
+        .mappings()
+        .all()
+    )
+
+
+def _get_user_initiated_chats_in_past_time_interval(session, user_id) -> list[dict]:
+    return (
+        session.execute(
+            select(Conversation.created, GroupChat.is_dm)
+            .join(Conversation, GroupChat.conversation_id == Conversation.id)
+            .where(GroupChat.creator_id == user_id)
+            .where(Conversation.created >= now() - RATE_LIMIT_INTERVAL)
+        )
+        .mappings()
+        .all()
+    )
+
+
+def _get_user_events_in_past_time_interval(session, user_id: int, action: RateLimitAction) -> list[dict]:
+    """Get all events for the user in the last rate limit interval for the mod email."""
+    match action:
+        case RateLimitAction.host_request:
+            return _get_user_host_requests_in_past_time_interval(session, user_id)
+        case RateLimitAction.friend_request:
+            return _get_user_friend_requests_in_past_time_interval(session, user_id)
+        case RateLimitAction.chat_initiation:
+            return _get_user_initiated_chats_in_past_time_interval(session, user_id)
+        case _:
+            raise ValueError(f"Unknown rate limit action: {action}")
```

Comment:
I would just return *all* events in the email, regardless of the rate limit type that the user hit. That gives mods a better view.

Reply:
Was a bit afraid of overloading the moderators with excessive information, but yeah, probably makes sense to rather serve more info. Added it 👍